pkgs/applications/blockchains/nimbus/default.nix

```diff
@@ -0,0 +1,64 @@
+{
+  stdenv, fetchgit, lib, nim, which, writeScriptBin,
+  # Options: nimbus_light_client, nimbus_validator_client, nimbus_signing_node
+  makeTargets ? [ "nimbus_beacon_node" ],
+  # WARNING: CPU optmizations that make binary not portable.
+  nativeBuild ? false
+}:
+
+stdenv.mkDerivation rec {
+  pname = "nimbus";
+  version = "22.11.0";
+  commit = "356dd4ee";
+  name = "${pname}-${version}-${commit}";
+
+  src = fetchgit {
+    url = "https://github.com/status-im/nimbus-eth2.git";
+    rev = "v${version}";
+    sha256 = "sha256-vtXhxR8bLbPE5orItiz9dRVa0KrSg4XOQ4fhQfJCYaQ=";
+    fetchSubmodules = true;
+  };
+
+  buildInputs = let
+    # Fix for Nim compiler calling 'git rev-parse' and 'lsb_release'.
+    fakeGit = writeScriptBin "git" "echo $commit";
+    fakeLsbRelease = writeScriptBin "lsb_release" "echo nix";
+  in [ fakeGit fakeLsbRelease which nim ];
+
+  enableParallelBuilding = true;
+
+  NIMFLAGS = lib.optionalString (!nativeBuild) " -d:disableMarchNative";
+
+  # WARNING: Version 1.6.10 is known to be stable.
+  makeFlags = makeTargets ++ [ "USE_SYSTEM_NIM=1" ];
```

Comment:
this is likely to break - ie, we only support building nimbus with the exact version that is pinned in the nimbus submodule - any deviation from that typically results in a broken binary, so I'd strongly recommend against using the system nim

Reply:
Have you actually looked at the diff?
https://github.com/NixOS/nixpkgs/blob/f7a6372abec3e3658f3856f685617e5d91d1ffc8/pkgs/applications/blockchains/nimbus/default.nix#L19-L20
https://github.com/NixOS/nixpkgs/blob/f7a6372abec3e3658f3856f685617e5d91d1ffc8/pkgs/top-level/all-packages.nix#L33711-L33719
I lock it in `pkgs/top-level/all-packages.nix`.

Reply:
well, this holds true for _this_ particular version of nimbus, but it's not a sustainable way to maintain the package - ie occasionally we patch Nim itself and then checks like this will lose their meaning (it has to be the exact same hash as we used). This is also important to protect against supply chain issues, such as a compromised compiler that includes key loggers and other nasty things, as has happened in the past.

Nim is just "one more library" in the view of nimbus-eth2, just like libp2p or any other package in the "vendor" folder. It just happens to be the compiler, but we audit each upgrade just like anything else.

Reply:
And since I'll be maintaining this package I can also update the necessary compiler version.
Separating build process into as many separate derivations is a good practice, it simplifies the package derivation, avoids unnecessary rebuilds, and allows for use of NixOS binary cache. I think this is a good solution.

Reply:
> I think this is a good solution.

fair, as long as it's the status-im/Nim repo and equivalence is maintained - the right "assert" to make would be that the git hash from the separate derivation matches the one of the submodule in nimbus-eth2 so that changes to either are reflected correctly and not "accidentally" mismatched when someone edits the package files for a seemingly benign upgrade.

Reply:
I recommended to @jakubgs to follow this approach as it is indeed the Nix-friendly way to construct the package.

Reply:
sure - the approach is fine as long as the `assert` above is changed to use the hash - ie the objective is to ensure that the exact same nim version is used (ie comparing version number is borderline useless) - I imagine it would read the hash from the submodule and compare it against whatever was used to produce the nim package.

Reply:
Indeed, it does appear that `nim_builder` puts `nim` in `nativeBuildInputs`:
https://github.com/NixOS/nixpkgs/blob/19e4cd1f95fc6de85659776b9a65e84062ca74e7/pkgs/development/nim-packages/nim_builder/default.nix#L7
So that is probably a good idea in our case as well.

Reply:
As for asserting the hash, you mean the hash of the `src` added in `pkgs/top-level/all-packages.nix`? Sure, lets do that.

Reply:
Actually, I don't know if I can do that:
```
nix-repl> nim.src.sha256
error: attribute 'src' missing

       at «string»:1:1:

            1| nim.src.sha256
             | ^

nix-repl> nim.nim.src.sha256
error: attribute 'sha256' missing

       at «string»:1:1:

            1| nim.nim.src.sha256
             | ^
```

Reply:
I'm thinking comparing against the hash that the nimbus-eth2 has registered in its submodule:

```sh
git -C vendor/nimbus-build-system/vendor/Nim rev-parse HEAD
7f90bcf5b4cbe5b7da534dd79bafbb4cafb313fa
```

Reply:
ie the build of nim would have to store this in some package metadata then the nimbus build would have to reload it - _in theory_, Nim can print it but this is quite unreliable / could be spoofed by someone malicious (ie it would be better if the build system fetched this from git as part of the build process) - it's still a better version check than relying on version number..

```
[arnetheduck@praeceps nimbus-eth2]$ ./env.sh nim -v
Nim Compiler Version 1.6.11 [Linux: amd64]
Compiled at 2022-12-01
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: 7f90bcf5b4cbe5b7da534dd79bafbb4cafb313fa
active boot switches: -d:release
```

Reply:
NixOS is a distribution and we generally do not want to use vendored libaries because they might be out of date, miss security patches or distro specific patches like https://github.com/NixOS/nixpkgs/tree/master/pkgs/development/compilers/nim